File: Som/LearningProcessor/SomLearningProcessor.cs
﻿using System;
using System.Collections.Generic;
using Som.Data.Suffle;
using Som.Learning;
using Som.Data;
using Som.Metrics;
using Som.Network;
using Som.Topology;

namespace Som.LearningProcessor
{
    public interface ILearningProcessor
    {
        void Learn();
    }

    public class SomLearningProcessor : ILearningProcessor
    {
        public INetwork Network { get; private set; }
        public ITopology Topology { get; private set; }

        protected IShuffleProvider ShuffleProvider { get; set; }
        protected ILearningDataProvider LearningDataProvider { get; set; }
        protected IRadiusProvider RadiusProvider { get; private set; }
        protected INeighbourhoodFunction NeighbourhoodFunction { get; set; }
        protected IMetricFunction MetricFunction { get; set; }
        protected ILearningFactorFunction LearningFactorFunction { get; set; }
        
        public int MaxIterationsCount { get; protected set; }

        public SomLearningProcessor(
            ILearningDataProvider learningDataProvider,
            INetwork network,
            IMetricFunction metricFunction,
            ILearningFactorFunction learningFactorFunction,
            INeighbourhoodFunction neighbourhoodFunction,
            int maxIterationsCount,
            IShuffleProvider shuffleProvider)
        {
            LearningDataProvider = learningDataProvider;
            Network = network;
            Topology = network.Topology;
            MetricFunction = metricFunction;
            LearningFactorFunction = learningFactorFunction;
            NeighbourhoodFunction = neighbourhoodFunction;
            MaxIterationsCount = maxIterationsCount;
            ShuffleProvider = shuffleProvider;

            RadiusProvider = new DefaultRadiusProvider(maxIterationsCount, Topology.WholeTopologyRadius);
        }

        public virtual void Learn()
        {
            int vectorsCount = LearningDataProvider.LearningVectorsCount;
            IList<int> suffleList = new ShuffleList(vectorsCount);

            for (int iteration = 0; iteration < MaxIterationsCount; iteration++)
            {
                //if ((iteration % 1000) == 0) Console.Write(string.Format("{0} ", iteration));
                suffleList = ShuffleProvider.Suffle(suffleList);

                for (int dataInd = 0; dataInd < vectorsCount; dataInd++)
                {
                    double[] dataVector = LearningDataProvider.GetLearingDataVector(suffleList[dataInd]);

                    int bestNeuronNum = FindBestMatchingNeuron(dataVector);

                    AccommodateNetworkWeights(bestNeuronNum, dataVector, iteration);
                }
            }
        }

        public virtual int FindBestMatchingNeuron(double[] dataVector)
        {
            int result = -1;
            Double minDistance = Double.MaxValue;
            for (int i = 0; i < Network.Neurons.Count; i++)
            {
                double distance = MetricFunction.GetDistance(Network.Neurons[i].Weights, dataVector);
                if (distance < minDistance)
                {
                    minDistance = distance;
                    result = i;
                }
            }
            return result;
        }

        public virtual void AccommodateNetworkWeights(int bestNeuronNum, double[] dataVector, int iteration)
        {
            var radius = RadiusProvider.GetRadius(iteration);
            var effectedNeurons = Topology.GetNeuronsInRadius(bestNeuronNum, radius);

            foreach (var effectedNeuron in effectedNeurons.Keys)
            {
                var distance = effectedNeurons[effectedNeuron];

                AccommodateNeuronWeights(effectedNeuron, dataVector, iteration, distance, radius);
            }
        }

        protected virtual void AccommodateNeuronWeights(int neuronNumber, double[] dataVector, int iteration, double distance, double radius)
        {
            var neuronWghts = Network.Neurons[neuronNumber].Weights;

            var learningRate = LearningFactorFunction.GetLearningRate(iteration);
            var falloffRate = NeighbourhoodFunction.GetDistanceFalloff(distance, radius);

            for (int i = 0; i < neuronWghts.Length; i++)
            {
                double weight = neuronWghts[i];
                neuronWghts[i] += learningRate * falloffRate * (dataVector[i] - weight);
            }
        }
    }

}

File: JavaSom/LearningProcessor/SomLearningProcessor.java
package LearningProcessor;

import Data.Shuffle.*;
import Learning.*;
import Data.*;
import Metrics.*;
import Network.*;
import Topology.*;

	public class SomLearningProcessor implements ILearningProcessor
	{
		private INetwork privateNetwork;
		public final INetwork getNetwork()
		{
			return privateNetwork;
		}
		private void setNetwork(INetwork value)
		{
			privateNetwork = value;
		}
		private ITopology privateTopology;
		public final ITopology getTopology()
		{
			return privateTopology;
		}
		private void setTopology(ITopology value)
		{
			privateTopology = value;
		}

		private IShuffleProvider privateShuffleProvider;
		protected final IShuffleProvider getShuffleProvider()
		{
			return privateShuffleProvider;
		}
		protected final void setShuffleProvider(IShuffleProvider value)
		{
			privateShuffleProvider = value;
		}
		private ILearningDataProvider privateLearningDataProvider;
		protected final ILearningDataProvider getLearningDataProvider()
		{
			return privateLearningDataProvider;
		}
		protected final void setLearningDataProvider(ILearningDataProvider value)
		{
			privateLearningDataProvider = value;
		}
		private IRadiusProvider privateRadiusProvider;
		protected final IRadiusProvider getRadiusProvider()
		{
			return privateRadiusProvider;
		}
		private void setRadiusProvider(IRadiusProvider value)
		{
			privateRadiusProvider = value;
		}
		private INeighbourhoodFunction privateNeighbourhoodFunction;
		protected final INeighbourhoodFunction getNeighbourhoodFunction()
		{
			return privateNeighbourhoodFunction;
		}
		protected final void setNeighbourhoodFunction(INeighbourhoodFunction value)
		{
			privateNeighbourhoodFunction = value;
		}
		private IMetricFunction privateMetricFunction;
		protected final IMetricFunction getMetricFunction()
		{
			return privateMetricFunction;
		}
		protected final void setMetricFunction(IMetricFunction value)
		{
			privateMetricFunction = value;
		}
		private ILearningFactorFunction privateLearningFactorFunction;
		protected final ILearningFactorFunction getLearningFactorFunction()
		{
			return privateLearningFactorFunction;
		}
		protected final void setLearningFactorFunction(ILearningFactorFunction value)
		{
			privateLearningFactorFunction = value;
		}

		private int privateMaxIterationsCount;
		public final int getMaxIterationsCount()
		{
			return privateMaxIterationsCount;
		}
		protected final void setMaxIterationsCount(int value)
		{
			privateMaxIterationsCount = value;
		}

		public SomLearningProcessor(ILearningDataProvider learningDataProvider, INetwork network, IMetricFunction metricFunction, ILearningFactorFunction learningFactorFunction, INeighbourhoodFunction neighbourhoodFunction, int maxIterationsCount, IShuffleProvider shuffleProvider)
		{
			setLearningDataProvider(learningDataProvider);
			setNetwork(network);
			setTopology(network.getTopology());
			setMetricFunction(metricFunction);
			setLearningFactorFunction(learningFactorFunction);
			setNeighbourhoodFunction(neighbourhoodFunction);
			setMaxIterationsCount(maxIterationsCount);
			setShuffleProvider(shuffleProvider);

			setRadiusProvider(new DefaultRadiusProvider(maxIterationsCount, getTopology().getWholeTopologyRadius()));
		}

		public void Learn()
		{
			int vectorsCount = getLearningDataProvider().getLearningVectorsCount();
			java.util.List<Integer> suffleList = new ShuffleList(vectorsCount);

			for (int iteration = 0; iteration < getMaxIterationsCount(); iteration++)
			{
				//if ((iteration % 1000) == 0) Console.Write(string.Format("{0} ", iteration));
				suffleList = getShuffleProvider().Suffle(suffleList);

				for (int dataInd = 0; dataInd < vectorsCount; dataInd++)
				{
					double[] dataVector = getLearningDataProvider().GetLearingDataVector(suffleList.get(dataInd));

					int bestNeuronNum = FindBestMatchingNeuron(dataVector);

					AccommodateNetworkWeights(bestNeuronNum, dataVector, iteration);
				}
			}
		}

		public int FindBestMatchingNeuron(double[] dataVector)
		{
			int result = -1;
			double minDistance = Double.MAX_VALUE;
			for (int i = 0; i < getNetwork().getNeurons().size(); i++)
			{
				double distance = getMetricFunction().GetDistance(getNetwork().getNeurons().get(i).getWeights(), dataVector);
				if (distance < minDistance)
				{
					minDistance = distance;
					result = i;
				}
			}
			return result;
		}

		public void AccommodateNetworkWeights(int bestNeuronNum, double[] dataVector, int iteration)
		{
			double radius = getRadiusProvider().GetRadius(iteration);
			java.util.HashMap<Integer, Double> effectedNeurons = getTopology().GetNeuronsInRadius(bestNeuronNum, radius);

			for (int effectedNeuron : effectedNeurons.keySet())
			{
				double distance = effectedNeurons.get(effectedNeuron);

				AccommodateNeuronWeights(effectedNeuron, dataVector, iteration, distance, radius);
			}
		}

		protected void AccommodateNeuronWeights(int neuronNumber, double[] dataVector, int iteration, double distance, double radius)
		{
			double[] neuronWghts = getNetwork().getNeurons().get(neuronNumber).getWeights();

			double learningRate = getLearningFactorFunction().GetLearningRate(iteration);
			double falloffRate = getNeighbourhoodFunction().GetDistanceFalloff(distance, radius);

			for (int i = 0; i < neuronWghts.length; i++)
			{
				double weight = neuronWghts[i];
				neuronWghts[i] += learningRate * falloffRate * (dataVector[i] - weight);
			}
		}
	}